README.md
./front/README.md

CHANGELOG.md
# simple-bug-reporter

## 1.1.2

- Remove useless Redux dependencies

## 1.1.1

- Update README.md

## 1.1.0

- Add a "data" props to pass any object to write into the email report
- **BREAKING CHANGES**
    - Remove BugReporterRedux: use BugReporter with the "data" props now
    - Rename "reduxStore" to "data" into MongoDB

## 1.0.5

- Fix CSS error

## 1.0.4

- Make the server part finally works...
- Update dependencies
- Move credentials to a .env file
- Clean code

## 1.0.3

- Update axios dependency to fix a security vulnerability

## 1.0.2

- Add English translations
- Add a "lang" option to choose between French and English

## 1.0.1

- Update README.md

## 1.0.0

- Initial release


server/server.js
const dotenv = require('dotenv');
dotenv.config();

const express = require('express');
const bodyParser = require('body-parser');
const helmet = require('helmet');
const mongoose = require('mongoose');
const nodemailer = require('nodemailer');

if (!process.env.MONGODB_URI
  || !process.env.RECEIVER_MAIL_ADDRESS
  || !process.env.SENDER_MAIL_SMTP_SERVER
  || !process.env.SENDER_MAIL_SMTP_PORT
  || !process.env.SENDER_MAIL_SECURE
  || !process.env.SENDER_MAIL_USER
  || !process.env.SENDER_MAIL_PASSWORD) {
  throw new Error('Missing environment variable(s).')
}

mongoose.connect(process.env.MONGODB_URI, { useNewUrlParser: true }, (err) => {
  if (err) { throw err }
})

const reportSchema = new mongoose.Schema({
  clientName: String,
  message: String,
  screenshot: String,
  data: String,
  createdAt: { type: Date, default: Date.now }
})

const ReportModel = mongoose.model('reports', reportSchema)

const app = express()
app.set('port', process.env.PORT || 4000)

app.use(bodyParser.json({ limit: '10mb' }))
app.use(helmet())

const allowCrossDomain = (req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*')
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS')
  res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type')
  res.setHeader('Access-Control-Allow-Credentials', true)

  if (req.method === 'OPTIONS') {
    return res.sendStatus(204)
  } else {
    next()
  }
}

app.use(allowCrossDomain)

const transporter = nodemailer.createTransport({
  host: process.env.SENDER_MAIL_SMTP_SERVER,
  port:  process.env.SENDER_MAIL_SMTP_PORT,
  secureConnection: process.env.SENDER_MAIL_SECURE,
  auth: {
    user: process.env.SENDER_MAIL_USER,
    pass: process.env.SENDER_MAIL_PASSWORD
  }
})

app.post('/', (req, res) => {
  const report = new ReportModel({
    clientName: req.body.clientName,
    message: req.body.message,
    data: req.body.data
  })

  if (req.body.screenshot) {
    report.screenshot = req.body.screenshot
  }

  report.save((err, savedReport) => {
    if (err) {
      return res.status(500).end('Error while saving data into the MongoDB database.')
    }

    console.log(`[${savedReport.createdAt}] Report from ${savedReport.clientName} saved.`)
  })

  const mailOptions = {
    from: 'bug-reporter@node.com',
    to: process.env.RECEIVER_MAIL_ADDRESS,
    subject: `${report.clientName} - Bug reported`,
    html: `
      <h3>Bug reported for "${report.clientName}"</h3>
      <p>Message: ${report.message}</p>
      <p>Data : ${report.data}</p>
      <img src="cid:screenshot@bug-reporter.com" alt="Screenshot"/>
    `,
    attachments: [{
      filename: 'screenshot.jpeg',
      path: report.screenshot,
      cid: 'screenshot@bug-reporter.com'
    }]
  }

  transporter.sendMail(mailOptions, (err) => {
    if (err) {
      console.log('ERROR', err)
    } else {
      console.log('Mail sent')
    }
  })

  return res.sendStatus(201)
})

app.all('*', (req, res) => {
  return res.sendStatus(404)
})

app.listen(app.get('port'), () => {
  console.log(`Bug reporter server on port #${app.get('port')}`) // eslint-disable-line no-console
})

module.exports = app;


front/webpack.config.js
const autoprefixer = require('autoprefixer');
const path = require('path');

module.exports = {
  entry: [
    require.resolve('babel-regenerator-runtime'),
    './src/index.js'
  ],
  output: {
    path: path.resolve(__dirname, 'build'),
    filename: 'index.js',
    libraryTarget: 'commonjs2'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        include: path.resolve(__dirname, 'src'),
        exclude: /(node_modules|build)/,
        use: {
          loader: require.resolve('babel-loader'),
          options: {
            presets: ['env']
          }
        }
      },
      {
        test: /\.css$/,
        use: [
          require.resolve('style-loader'),
          {
            loader: require.resolve('css-loader'),
            options: {
              importLoaders: 1,
            },
          },
          {
            loader: require.resolve('postcss-loader'),
            options: {
              ident: 'postcss',
              plugins: () => [
                require('postcss-flexbugs-fixes'),
                autoprefixer({
                  browsers: [
                    '>1%',
                    'last 2 versions',
                    'Firefox ESR',
                    'not ie < 9', // React doesn't support IE8 anyway
                  ],
                  flexbox: 'no-2009',
                }),
              ],
            },
          },
        ],
      }
    ]
  },
  externals: {
    'react': 'commonjs react'
  }
};


front/src/BugReporter.js
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import html2canvas from 'html2canvas';
import axios from 'axios';

import translations from './translations.json';
import './bug-reporter.css';

class BugReporter extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      message: '',
      screenshotChecked: true,
      selectElementMode: false,
      element: null,
      loading: false
    };

    this.onMessageChange = this.onMessageChange.bind(this);
    this.onScreenshotCheck = this.onScreenshotCheck.bind(this);
    this.registerClick = this.registerClick.bind(this);
    this.setSelectElementMode = this.setSelectElementMode.bind(this);
    this.submit = this.submit.bind(this);
    this.cancel = this.cancel.bind(this);
  }

  onMessageChange(event) {
    this.setState({
      message: event.target.value
    });
  }

  onScreenshotCheck(event) {
    this.setState({
      screenshotChecked: event.target.checked
    });
  }

  setSelectElementMode() {
    this.setState({
      selectElementMode: true
    });
  }

  registerClick(event) {
    this.setState({
      selectElementMode: false,
      element: {
        x: event.nativeEvent.clientX,
        y: event.nativeEvent.clientY
      }
    });
  }

  translate(name) {
    if (translations[this.props.lang] === undefined) {
      return translations.en[name];
    }

    return translations[this.props.lang][name];
  }

  async submit() {
    const { message, element } = this.state;
    let dataScreenshot;

    this.setState({
      loading: true
    });

    try {
      if (this.state.screenshotChecked) {
        const canvas = await html2canvas(document.body);
        if (element) {
          const context = canvas.getContext('2d');
          context.beginPath();
          context.arc(element.x, element.y, this.props.annotationRadius, 0, 2 * Math.PI, false);
          context.fillStyle = this.props.annotationColor;
          context.fill();
        }

        dataScreenshot = canvas.toDataURL('image/jpeg', this.props.screenshotQuality);
      }

      await axios.post(this.props.serverURL, {
        clientName: this.props.name,
        message,
        screenshot: dataScreenshot,
        data: JSON.stringify(this.props.data)
      });

      alert(this.translate('report_sent'));

    } catch (err) {
      console.error(err);
      alert(this.translate('error_sending'));
    }

    this.setState({
      message: '',
      screenshotChecked: true,
      element: null,
      loading: false
    });
  }

  cancel() {
    this.setState({
      message: '',
      screenshotChecked: true,
      element: null
    });
  }

  render() {
    if (!this.props.dev) {
      return null;
    }

    return (
      <div id="bug-reporter" className={classNames({ 'bug-reporter-loader': this.state.loading })}>
        <div className="title">
          Bug reporter
        </div>

        <div className="field-group message-input">
          <label htmlFor="message">Message</label>
          <textarea
            id="message"
            name="message"
            rows="1"
            value={this.state.message}
            onChange={this.onMessageChange}
          />
        </div>

        <div className="field-group screenshot-input">
          <label htmlFor="screenshot">
            <input
              type="checkbox"
              id="screenshot"
              name="screenshot"
              disabled={this.state.message === ''}
              checked={this.state.screenshotChecked}
              onChange={this.onScreenshotCheck}
            />
            {this.translate('screenshot_button')}
          </label>
        </div>

        <div className="field-group select-input">
          <button
            type="button"
            disabled={this.state.message === '' || !this.state.screenshotChecked}
            onClick={this.setSelectElementMode}
          >
            { this.state.element ? this.translate('set_element_button') : this.translate('select_element_button') }
          </button>
        </div>

        <div className="buttons-group">
          <div className="field-group submit-input">
            <button
              type="button"
              disabled={this.state.message === ''}
              onClick={this.submit}
            >
              {this.translate('send_button')}
            </button>
          </div>

          <div className="field-group cancel-input">
            <button
              type="button"
              onClick={this.cancel}
            >
              {this.translate('cancel_button')}
            </button>
          </div>
        </div>

        <div
          className={classNames('overlay', { hidden: !this.state.selectElementMode })}
          onClick={this.registerClick}
          role="presentation"
        >
          <div className="overlay-text">
            {this.translate('select_element_message')}
          </div>
        </div>
      </div>
    );
  }

}

BugReporter.defaultProps = {
  annotationColor: 'rgba(255, 0, 0, 0.7)',
  annotationRadius: 5,
  data: {},
  dev: true,
  lang: 'en',
  screenshotQuality: 0.6
};

BugReporter.propTypes = {
  annotationColor: PropTypes.string,
  annotationRadius: PropTypes.number,
  data: PropTypes.object,
  dev: PropTypes.bool,
  lang: PropTypes.string,
  name: PropTypes.string.isRequired,
  screenshotQuality: PropTypes.number,
  serverURL: PropTypes.string.isRequired
};

export default BugReporter;


front/src/BugReporterRedux.js
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import html2canvas from 'html2canvas';
import axios from 'axios';

import translations from './translations.json';
import './bug-reporter.css';

class BugReporter extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      message: '',
      screenshotChecked: true,
      selectElementMode: false,
      element: null,
      loading: false
    };

    this.onMessageChange = this.onMessageChange.bind(this);
    this.onScreenshotCheck = this.onScreenshotCheck.bind(this);
    this.registerClick = this.registerClick.bind(this);
    this.setSelectElementMode = this.setSelectElementMode.bind(this);
    this.submit = this.submit.bind(this);
    this.cancel = this.cancel.bind(this);
  }

  onMessageChange(event) {
    this.setState({
      message: event.target.value
    });
  }

  onScreenshotCheck(event) {
    this.setState({
      screenshotChecked: event.target.checked
    });
  }

  setSelectElementMode() {
    this.setState({
      selectElementMode: true
    });
  }

  registerClick(event) {
    this.setState({
      selectElementMode: false,
      element: {
        x: event.nativeEvent.clientX,
        y: event.nativeEvent.clientY
      }
    });
  }

  translate(name) {
    if (translations[this.props.lang] === undefined) {
      return translations.en[name];
    }

    return translations[this.props.lang][name];
  }

  async submit() {
    const { message, element } = this.state;
    let dataScreenshot;

    this.setState({
      loading: true
    });

    try {
      if (this.state.screenshotChecked) {
        const canvas = await html2canvas(document.body);
        if (element) {
          const context = canvas.getContext('2d');
          context.beginPath();
          context.arc(element.x, element.y, this.props.annotationRadius, 0, 2 * Math.PI, false);
          context.fillStyle = this.props.annotationColor;
          context.fill();
        }

        dataScreenshot = canvas.toDataURL('image/jpeg', this.props.screenshotQuality);
      }

      const reduxStore = JSON.stringify(this.props.store);

      await axios.post(this.props.serverURL, {
        clientName: this.props.name,
        message,
        screenshot: dataScreenshot,
        reduxStore
      });

      alert(this.translate('report_sent'));

    } catch (err) {
      console.error(err);
      alert(this.translate('error_sending'));
    }

    this.setState({
      message: '',
      screenshotChecked: true,
      element: null,
      loading: false
    });
  }

  cancel() {
    this.setState({
      message: '',
      screenshotChecked: true,
      element: null
    });
  }

  render() {
    if (!this.props.dev) {
      return null;
    }

    return (
      <div id="bug-reporter" className={classNames({ 'bug-reporter-loader': this.state.loading })}>
        <div className="title">
          Bug reporter
        </div>

        <div className="field-group message-input">
          <label htmlFor="message">
            Message
            <textarea
              id="message"
              name="message"
              rows="1"
              value={this.state.message}
              onChange={this.onMessageChange}
            />
          </label>
        </div>

        <div className="field-group screenshot-input">
          <label htmlFor="screenshot">
            <input
              type="checkbox"
              id="screenshot"
              name="screenshot"
              disabled={this.state.message === ''}
              checked={this.state.screenshotChecked}
              onChange={this.onScreenshotCheck}
            />
            {this.translate('screenshot_button')}
          </label>
        </div>

        <div className="field-group select-input">
          <button
            type="button"
            disabled={this.state.message === '' || !this.state.screenshotChecked}
            onClick={this.setSelectElementMode}
          >
            { this.state.element ? this.translate('set_element_button') : this.translate('select_element_button') }
          </button>
        </div>

        <div className="buttons-group">
          <div className="field-group submit-input">
            <button
              type="button"
              disabled={this.state.message === ''}
              onClick={this.submit}
            >
              {this.translate('send_button')}
            </button>
          </div>

          <div className="field-group cancel-input">
            <button
              type="button"
              onClick={this.cancel}
            >
              {this.translate('cancel_button')}
            </button>
          </div>
        </div>

        <div
          className={classNames('overlay', { hidden: !this.state.selectElementMode })}
          onClick={this.registerClick}
          role="presentation"
        >
          <div className="overlay-text">
            {this.translate('select_element_message')}
          </div>
        </div>
      </div>
    );
  }

}

BugReporter.defaultProps = {
  annotationColor: 'rgba(255, 0, 0, 0.7)',
  annotationRadius: 5,
  dev: true,
  lang: 'en',
  screenshotQuality: 0.6
};

BugReporter.propTypes = {
  annotationColor: PropTypes.string,
  annotationRadius: PropTypes.number,
  dev: PropTypes.bool,
  lang: PropTypes.string,
  name: PropTypes.string.isRequired,
  screenshotQuality: PropTypes.number,
  serverURL: PropTypes.string.isRequired,
  store: PropTypes.object.isRequired
};

function mapStateToProps(state) {
  return {
    store: state
  };
}

export default connect(mapStateToProps)(BugReporter);
